gapis/memory/decoder.go

```diff
@@ -0,0 +1,152 @@
+// Copyright (C) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memory
+
+import (
+	"github.com/google/gapid/core/data/binary"
+	"github.com/google/gapid/core/math/u64"
+	"github.com/google/gapid/core/os/device"
+)
+
+// Decoder provides methods to read primitives from a binary.Reader, respecting
+// a given MemoryLayout.
+// Decoder will automatically handle alignment and types sizes.
+type Decoder struct {
+	r binary.Reader
+	m *device.MemoryLayout
+	o uint64
```

Comment:
So this field just shadows the offset of the Reader right?  I always find it a bit sad to see copied state.  I worry they might go out of sync.  Oh well...